src/System.Net.Http/tests/UnitTests/HPack/DynamicTableTest.cs

```diff
@@ -55,5 +57,57 @@ void InsertOne()
                 Assert.True(expectedData.AsSpan().SequenceEqual(dynamicField.Value));
             }
         }
+
+        [Theory]
+        [MemberData(nameof(CreateResizeData))]
+        public void DynamicTable_Resize_Success(int initialMaxSize, int finalMaxSize, int insertSize)
+        {
+            // This is purely to make it simple to perfectly reach our initial max size to test growing a full but non-wrapping buffer.
+            Debug.Assert((insertSize % 64) == 0, $"{nameof(insertSize)} must be a multiple of 64 ({nameof(HeaderField)}.{nameof(HeaderField.RfcOverhead)} * 2)");
+
+            var dynamicTable = new DynamicTable(maxSize: initialMaxSize);
+            int insertedSize = 0;
+
+            while (insertedSize != insertSize)
+            {
+                byte[] data = Encoding.ASCII.GetBytes($"header-{dynamicTable.Size}".PadRight(16, ' '));
+                Debug.Assert(data.Length == 16);
+
+                dynamicTable.Insert(data, data);
+                insertedSize += data.Length * 2 + HeaderField.RfcOverhead;
+            }
+
+            List<HeaderField> headers = new List<HeaderField>();
+
+            for (int i = 0; i < dynamicTable.Count; ++i)
+            {
+                headers.Add(dynamicTable[i]);
+            }
+
+            dynamicTable.Resize(finalMaxSize);
+
+            int expectedCount = Math.Min(finalMaxSize / 64, headers.Count);
+            Assert.Equal(expectedCount, dynamicTable.Count);
+
+            for (int i = 0; i < dynamicTable.Count; ++i)
+            {
+                Assert.True(headers[i].Name.AsSpan().SequenceEqual(dynamicTable[i].Name));
+                Assert.True(headers[i].Value.AsSpan().SequenceEqual(dynamicTable[i].Value));
+            }
+        }
+
+        public static IEnumerable<object[]> CreateResizeData()
+        {
+            for (int initialMaxSize = 128; initialMaxSize <= 512; initialMaxSize += 128)
+            {
+                for (int finalMaxSize = 128; finalMaxSize <= 512; finalMaxSize += 128)
+                {
+                    for (int insertSize = 128; insertSize <= 512; insertSize += 128)
+                    {
+                        yield return new object[] { initialMaxSize, finalMaxSize, insertSize };
+                    }
+                }
+            }
```

Comment:
FWIW, this might be simpler as:
```C#
public static IEnumerable<object[]> CreateResizeData()
{
    var values = new[] { 128, 256, 384, 512 };
    return from initialMaxSize in values
           from finalMaxSize in values
           from insertSize in values
           select new object[] { initialMaxSize, finalMaxSize, insertSize };
}
```